fairscale/nn/wrap/auto_wrap.py

```diff
@@ -130,68 +175,85 @@ class ConfigAutoWrap:
+    in_autowrap_context: bool = False  # Context flag
+    wrapper_cls: Optional[Callable] = None  # The wrapper class
+    kwargs: Dict[str, Any] = {}  # Wrapper's args
+    should_wrap: Optional[Callable] = None  # Used only in auto_wrap
 
-    def __init__(self, module_blocklist: Optional[List] = None, **kwargs: Dict[str, Any]):
-        if module_blocklist:
-            self.module_blocklist += module_blocklist
+    def __init__(self, should_wrap: Optional[Callable] = None, **kwargs: Dict[str, Any]):
+        self.should_wrap = should_wrap
         self.kwargs = kwargs
 
     @staticmethod
-    def enable_autowrap_context(kwargs: Any) -> None:
+    def enable_autowrap_context(should_wrap: Optional[Callable], kwargs: Any) -> None:
         if ConfigAutoWrap.in_autowrap_context:
             raise NotImplementedError(
                 "You are already within an autowrap context and we currently do not supported nested autowrap."
             )
         ConfigAutoWrap.in_autowrap_context = True
+        # Get and save the wrapper cls for the context.
+        assert "wrapper_cls" in kwargs.keys()
+        ConfigAutoWrap.wrapper_cls = cast(Callable, kwargs["wrapper_cls"])
+        del kwargs["wrapper_cls"]
+        # Save the rest.
+        ConfigAutoWrap.should_wrap = should_wrap_default if should_wrap is None else should_wrap
         ConfigAutoWrap.kwargs = kwargs
-        ConfigAutoWrap.module_blocklist += FSDP_MODULE_BLOCKLIST
 
     @staticmethod
     def disable_autowrap_context() -> None:
         ConfigAutoWrap.in_autowrap_context = False
+        ConfigAutoWrap.wrapper_cls = None
         ConfigAutoWrap.kwargs = {}
-        ConfigAutoWrap.module_blocklist = []
+        ConfigAutoWrap.should_wrap = None
 
     def __enter__(self) -> None:
-        self.enable_autowrap_context(self.kwargs)
+        self.enable_autowrap_context(self.should_wrap, self.kwargs)
 
     def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
         self.disable_autowrap_context()
 
     @staticmethod
-    def recursive_wrap(module: nn.Module, min_num_params: int, **kwargs: Any) -> Tuple[nn.Module, int]:
+    def recursive_wrap(module: nn.Module, should_wrap: Optional[Callable], **kwargs: Any) -> Tuple[nn.Module, int]:
         """
         Automatically wrap child modules of *module* that meet the given
         criteria with :func:`auto_wrap`.
 
         Args:
-            module (nn.Module): module to recursively wrap
-            min_num_params (int): min number of parameters for a child Module to
-                be wrapped
+            module (nn.Module):
+                module to recursively wrap
+            should_wrap (Callable, Optional):
+                optionally, override the :func:`should_wrap` from the context.
+
+        Returns:
+            (nn.Module, int):
+                Wrapped module and the number parameters wrapped recursively.
         """
-        if isinstance(module, tuple(ConfigAutoWrap.module_blocklist)):
```

Comment:
this if is not needed since the first should_wrap decision takes care of it now.